Windows build fails with compiler errors

When I attempt to build on Windows, I get the following error messages.

<img width="1194" height="572" alt="Image" src="https://github.com/user-attachments/assets/e240e281-29ef-4fa0-945e-fd8d2d71f083" />

These are real compiler errors, not warnings. I'm not sure why the macOS builds work. I am using VS2022 on Windows 11.

It appears to be my version of MSVC that has a bug. If the GitHub Windows CI is building, the suggests mitigation may not or may not be worth doing. Here is what Codex had to say:

### LLM Analysis

This appears to be an MSVC 14.44 conformance bug, not a unity-build issue.

  The optional-member macros in `OptionalMembers.h` use dependent expressions such as `decltype(el->endLength())` inside a C++20 `requires` expression. When instantiated for a type without that member, the requirement should evaluate to `false`, but MSVC 14.44 emits C2039 instead:

  ```text
  LineFunctions.h(83): error C2039: 'endLength': is not a member of 'mx::core::Wedge'
  FontFunctions.h(147): error C2039: 'lineThrough': is not a member of 'mx::core::EmptyFont'
```

I compiled DirectionWriter.cpp, LayoutFunctions.cpp, and NotationsWriter.cpp individually and reproduced the same errors, confirming that unity builds are not the cause.

This can be mitigated by checking whether the getter exists before deriving its optional `value_type`. Only form the dependent `decltype` and invoke the setter inside that successful check. This avoids MSVC evaluating the missing member as part of the more complex `requires` expression.

--- a/src/private/mx/utility/OptionalMembers.h
+++ b/src/private/mx/utility/OptionalMembers.h
@@ -101,22 +101,22 @@
                                                                                                                        \
     template <class T> bool lookForAndSetHas##fieldNameCapitalized(bool has, T *const el)                              \
     {                                                                                                                  \
-        if constexpr (requires { el->setterName(std::nullopt); })                                                      \
+        if constexpr (requires { el->fieldName(); })                                                                   \
         {                                                                                                              \
-            if (!has)                                                                                                  \
+            if constexpr (requires { el->setterName(std::nullopt); })                                                  \
             {                                                                                                          \
-                el->setterName(std::nullopt);                                                                          \
+                if (!has)                                                                                              \
+                {                                                                                                      \
+                    el->setterName(std::nullopt);                                                                      \
+                }                                                                                                      \
+                else if (!el->fieldName().has_value())                                                                 \
+                {                                                                                                      \
+                    el->setterName(typename std::decay_t<decltype(el->fieldName())>::value_type{});                    \
+                }                                                                                                      \
+                return has;                                                                                            \
             }                                                                                                          \
-            else if (!el->fieldName().has_value())                                                                     \
-            {                                                                                                          \
-                el->setterName(typename std::decay_t<decltype(el->fieldName())>::value_type{});                        \
-            }                                                                                                          \
-            return has;                                                                                                \
         }                                                                                                              \
-        else                                                                                                           \
-        {                                                                                                              \
-            return false;                                                                                              \
-        }                                                                                                              \
+        return false;                                                                                                  \
     }
 
 // Value setter for a clamped-decimal wrapper field.
@@ -124,15 +124,18 @@
                                                                                                                        \
     template <class T, class VALUE_TYPE> void lookForAndSet##fieldNameCapitalized(VALUE_TYPE value, T *const el)       \
     {                                                                                                                  \
-        if constexpr (requires {                                                                                       \
-                          el->setterName(typename std::decay_t<decltype(el->fieldName())>::value_type{                 \
-                              mx::core::Decimal{static_cast<double>(value)}});                                         \
-                      })                                                                                               \
+        if constexpr (requires { el->fieldName(); })                                                                   \
         {                                                                                                              \
-            if (el->fieldName().has_value())                                                                           \
+            if constexpr (requires {                                                                                   \
+                              el->setterName(typename std::decay_t<decltype(el->fieldName())>::value_type{             \
+                                  mx::core::Decimal{static_cast<double>(value)}});                                     \
+                          })                                                                                           \
             {                                                                                                          \
-                el->setterName(typename std::decay_t<decltype(el->fieldName())>::value_type{                           \
-                    mx::core::Decimal{static_cast<double>(value)}});                                                   \
+                if (el->fieldName().has_value())                                                                       \
+                {                                                                                                      \
+                    el->setterName(typename std::decay_t<decltype(el->fieldName())>::value_type{                       \
+                        mx::core::Decimal{static_cast<double>(value)}});                                               \
+                }                                                                                                      \
             }                                                                                                          \
         }                                                                                                              \
     }
@@ -142,13 +145,16 @@
                                                                                                                        \
     template <class T> void lookForAndSet##fieldNameCapitalized(int value, T *const el)                                \
     {                                                                                                                  \
-        if constexpr (requires {                                                                                       \
-                          el->setterName(typename std::decay_t<decltype(el->fieldName())>::value_type{value});         \
-                      })                                                                                               \
+        if constexpr (requires { el->fieldName(); })                                                                   \
         {                                                                                                              \
-            if (el->fieldName().has_value())                                                                           \
+            if constexpr (requires {                                                                                   \
+                              el->setterName(typename std::decay_t<decltype(el->fieldName())>::value_type{value});     \
+                          })                                                                                           \
             {                                                                                                          \
-                el->setterName(typename std::decay_t<decltype(el->fieldName())>::value_type{value});                   \
+                if (el->fieldName().has_value())                                                                       \
+                {                                                                                                      \
+                    el->setterName(typename std::decay_t<decltype(el->fieldName())>::value_type{value});               \
+                }                                                                                                      \
             }                                                                                                          \
         }                                                                                                              \
     }
@@ -158,11 +164,14 @@
                                                                                                                        \
     template <class T, class VALUE_TYPE> void lookForAndSet##fieldNameCapitalized(VALUE_TYPE value, T *const el)       \
     {                                                                                                                  \
-        if constexpr (requires { el->setterName(std::move(value)); })                                                  \
+        if constexpr (requires { el->fieldName(); })                                                                   \
         {                                                                                                              \
-            if (el->fieldName().has_value())                                                                           \
+            if constexpr (requires { el->setterName(std::move(value)); })                                              \
             {                                                                                                          \
-                el->setterName(std::move(value));                                                                      \
+                if (el->fieldName().has_value())                                                                       \
+                {                                                                                                      \
+                    el->setterName(std::move(value));                                                                  \
+                }                                                                                                      \
             }                                                                                                          \
         }                                                                                                              \
     }

